updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Result;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use log::warn;
+use std::{collections::HashMap, env, fs, path::Path, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
+
+type StateChangeCallback = Box<dyn Fn(State)>;
+type ErrorCallback = Box<dyn Fn(&str)>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    listeners: HashMap<StateID, Vec<StateChangeCallback>>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    pub id: StateID,
+    pub conn: UnixStream,
+}
+
+impl State {
+    // Cancels the current action on the agent.
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    // Tell the agent to proceed with the transition.
+    pub fn proceed(&self) {
+        // No message need to be sent to the connection in order to the
+        // agent to proceed handling the current state.
+    }
+
+    pub fn stop(&self) {
+        process::exit(1);
+    }
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on_state_change(&mut self, state: StateID, f: StateChangeCallback) {
+        match self.listeners.get_mut(&state) {
+            Some(callbacks) => {
+                callbacks.push(f);
+            }
+            None => {
+                self.listeners.insert(state, vec![Box::new(f)]);
+            }
+        }
+    }
+
+    pub fn on_error(&mut self, f: ErrorCallback) {
+        self.error_handlers.push(f);
+    }
+
+    pub async fn listen(&self) -> Result<()> {
+        if !Path::new(&SDK_TRIGGER_FILENAME).exists() {
+            warn!(
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}",
+                SDK_TRIGGER_FILENAME
+            );
+        }
+        let listener = match env::var("UH_LISTENER_TEST") {
+            Ok(ref uh_listener_test) => {
+                if Path::new(uh_listener_test).exists() {
+                    fs::remove_file(uh_listener_test)?;
+                }
+                UnixListener::bind(uh_listener_test).await?
+            }
+            _ => {
+                if Path::new(&SOCKET_PATH).exists() {
+                    fs::remove_file(SOCKET_PATH)?;
+                }
+                UnixListener::bind(SOCKET_PATH).await?
```

Comment:
There is duplicated logic here that can be easily avoided by metching only to extract the path string.